### PR TITLE
Add tf.linalg.gramSchmidt and tf.eye

### PR DESCRIPTION
**Pull request by caisq**
* `tf.linalg.gramSchmidt` performs Gram-Schmidt orthogonalization on
  Matrices or arrays of vectors. It will be used to speed up
  the orthogonal initialization of weights (e.g., in RNNs).
* `tf.eye` generates identity matrices.

Towards: https://github.com/tensorflow/tfjs/issues/245

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1024)

### CODE REVIEW

File: src/ops/linalg_ops.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    xs = xs as Tensor1D[];
+    assert(
+        xs.length <= xs[0].shape[0],
+        `Gram-Schmidt: Number of vectors (${xs.length}) exceeds ` +
+            `number of dimensions (${xs[0].shape[0]}).`);
+
+    const ys: Tensor1D[] = [];
+    for (let i = 0; i < xs.length; ++i) {
+      ys.push(Tracking.tidy(() => {
+        let x = (xs as Tensor1D[])[i];
```
Comment (bileschi) [inline src/ops/linalg_ops.ts]: `xs` was already cast to `Tensor1D` in line 62

File: src/ops/linalg_ops_test.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+import * as tf from '../index';
+import { describeWithFlags } from '../jasmine_util';
+import { Tensor1D, Tensor2D } from '../tensor';
+import { ALL_ENVS, expectArraysClose, WEBGL_ENVS } from '../test_util';
+
+describeWithFlags('gramSchmidt-tiny', ALL_ENVS, () => {
+  // TODO(cais): Small things should get tested on ALL_ENVS.
+  it('2x2, Array of Tensor1D', () => {
+    const xs: Tensor1D[] = [tf.tensor1d([1, 2]), tf.tensor1d([-1, 5])];
```
Comment (bileschi) [inline src/ops/linalg_ops_test.ts]: This might be a good candidate for a random array as input

File: src/ops/linalg_ops.ts (hunk 1 above)
Comment (caisq) [reply src/ops/linalg_ops.ts]: The linter complains if I remove this explicit casting.

File: src/ops/linalg_ops_test.ts (hunk 2 above)
Comment (caisq) [reply src/ops/linalg_ops_test.ts]: Done. Using random vectors and matrics in this test file, wherever it makes sense.

**Comment by bileschi** [comment]
:lgtm_strong:

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-:-LBzzHch-4kOjGiGfwJx:b-jbsf8z)*

**Comment by davidsoergel** [comment]
:lgtm_strong:

---

Reviewed 3 of 5 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[src/ops/array_ops.ts, line 376 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-LBzzLrT-qczaSMi5PZw:-LBzzLrT-qczaSMi5PZx:b-om2i3u) ([raw file](https://github.com/tensorflow/tfjs-core/blob/7d3c9d3074350dd967cd5a72668f67e524486168/src/ops/array_ops.ts#L376)):*
> ```
> Nnumber
> ```
typo

---

*[src/ops/linalg_ops.ts, line 71 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-LBv9F9UgzVmXoA2D8hd-r1-71:-LBzznnQ-fYEbUd6gSHw:b-ab5fqv) ([raw file](https://github.com/tensorflow/tfjs-core/blob/dccca2f77aecf0aefa4bfd15b385cd07e70711ab/src/ops/linalg_ops.ts#L71)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

The linter complains if I remove this explicit casting.
</blockquote></details>

Could avoid this by creating a local xs1d instead of overwriting the xs parameter

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-:-LC--E6L4-9Mc1He_9lg:b-jbsf8z)*

**Comment by caisq** [comment]
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/ops/linalg_ops.ts, line 71 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-LBv9F9UgzVmXoA2D8hd-r1-71:-LC-K-mxB3zRCunt08di:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/dccca2f77aecf0aefa4bfd15b385cd07e70711ab/src/ops/linalg_ops.ts#L71)):*
<details><summary><i>Previously, davidsoergel (David Soergel) wrote…</i></summary><blockquote>

Could avoid this by creating a local xs1d instead of overwriting the xs parameter
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1024)*

**Comment by caisq** [comment]
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[src/ops/array_ops.ts, line 376 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-LBzzLrT-qczaSMi5PZw:-LC-UNIfAKzLma7VsbZf:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/bcf12ed28072972ec0544707780eb1f0b1fa0e84/src/ops/array_ops.ts#L376)):*
<details><summary><i>Previously, davidsoergel (David Soergel) wrote…</i></summary><blockquote>

> ```
> Nnumber
> ```
typo
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1024)*

**Comment by dsmilkov** [comment]
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/ops/array_ops.ts, line 381 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-LC-TZ0o8xD97WlNolOX:-LC-TZ0o8xD97WlNolOY:b-3uznd8) ([raw file](https://github.com/tensorflow/tfjs-core/blob/c1c15e402f3438163534534c323af0e7c8a18d1e/src/ops/array_ops.ts#L381)):*
> ```XML
>    */
>   @operation
>   static eye(n: number, dtype: DataType = 'float32'): Tensor2D {
> ```

align with [tf.eye](https://www.tensorflow.org/api_docs/python/tf/eye) by adding `numRows`, `numColumns` and `batchShape`

---

*[src/ops/linalg_ops.ts, line 46 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-LC-UpJe4cv1QOoms-fk:-LC-UpJe4cv1QOoms-fl:bdtissm) ([raw file](https://github.com/tensorflow/tfjs-core/blob/9ff349c83367d74fedc8e78612950cb80bfcf4cb/src/ops/linalg_ops.ts#L46)):*
> ```XML
>    */
>   @operation
>   static gramSchmidt(xs: Tensor1D[]|Tensor2D): Tensor1D[]|Tensor2D {
> ```

what's the tensorflow's equivalent op and can we align that API?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1024)*

**Comment by caisq** [comment]
Review status: 1 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/ops/array_ops.ts, line 381 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-LC-TZ0o8xD97WlNolOX:-LC-oknS5gFcFkc6yZYE:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/c1c15e402f3438163534534c323af0e7c8a18d1e/src/ops/array_ops.ts#L381)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

align with [tf.eye](https://www.tensorflow.org/api_docs/python/tf/eye) by adding `numRows`, `numColumns` and `batchShape`
</blockquote></details>

Done.

---

*[src/ops/linalg_ops.ts, line 35 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-LC-XRN3-oqyZ25X9K5G:-LC-ot2h8Dy6Ag7YAlET:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/9ff349c83367d74fedc8e78612950cb80bfcf4cb/src/ops/linalg_ops.ts#L35)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Just say: - an array of `Tensor1D` (no need to mark Array in quotes and no need to mention `non-empty` since its implicit)
</blockquote></details>

Done.

---

*[src/ops/linalg_ops_test.ts, line 24 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1024#-LC-YTZ--lnAfJotMaEi:-LC-oxXtDx4sd0CK_U3W:b-gaify6) ([raw file](https://github.com/tensorflow/tfjs-core/blob/9ff349c83367d74fedc8e78612950cb80bfcf4cb/src/ops/linalg_ops_test.ts#L24)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

is this TODO outdated?
</blockquote></details>

Yes. Removed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1024)*

**Pull request merged**